src/metabase/xrays/automagic_dashboards/interesting.clj

```diff
@@ -100,43 +100,49 @@
   ->reference (fn [template-type model]
                 [template-type (mi/model model)]))
 
-(defn- optimal-datetime-resolution
+(defn- optimal-temporal-resolution
   [field]
   (let [[earliest latest] (some->> field
                                    :fingerprint
                                    :type
                                    :type/DateTime
                                    ((juxt :earliest :latest))
-                                   (map u.date/parse))]
+                                   (map u.date/parse))
+        date?             (isa? (:base_type field) :type/Date)
+        date-time?        (isa? (:base_type field) :type/DateTime)
+        time?             (isa? (:base_type field) :type/Time)]
     (if (and earliest latest)
-      ;; e.g. if 3 hours > [duration between earliest and latest] then use `:minute` resolution
-      (condp u.date/greater-than-period-duration? (u.date/period-duration earliest latest)
-        (t/hours 3) :minute
-        (t/days 7) :hour
-        (t/months 6) :day
-        (t/years 10) :month
-        :year)
-      :day)))
+      (let [duration      (u.date/period-duration earliest latest)
+            greater-than? #(u.date/greater-than-period-duration? % duration)]
+        (cond
+         ;; e.g. if 3 hours > [duration between earliest and latest] then use `:minute` resolution
+         (and (or time? date-time?) (greater-than? (t/hours 3)))  :minute
+         (and (or time? date-time?) (greater-than? (t/days 7)))   :hour
+         (and (or date? date-time?) (greater-than? (t/months 6))) :day
+         (and (or date? date-time?) (greater-than? (t/years 10))) :month
+         (or date? date-time?)                                    :year
+         time?                                                    :hour))
```

Comment:
This log is quite dense and tangled, and would be hard and failure prone to extend.

A more declarative approach could keep the old logic in a helper method, say `optimal-temporal-resolution*`, and use a data-driven approach to keep the values within a type determined range:

```clojure
(def time-resolutions [:minute :hour :day :month :year])

(def resolution-idx
  (into {} (map-indexed (fn [idx res] [res idx]) time-resolutions)))

(def type-constraints
  {:type/Time     [nil (resolution-idx :hour)]
   :type/Date     [(resolution-idx :day) nil]
   :type/DateTime [nil nil]})

(defn optimal-temporal-resolution [type field]
  (let [res               (optimal-temporal-resolution* field)
        res-idx           (resolution-idx res)
        [min-idx max-idx] (type-constraints type)
        constrained-idx   (cond-> res-idx
                            min-idx (max min-idx)
                            max-idx (min max-idx))]
    (time-resolutions constrained-idx)))
```

What do you think?

Reply:
tbh, I find that a bit complicated and rely on the order time resolutions.
I updated the code by using a helper to check whether it can use a resolution [4ac6d32](https://github.com/metabase/metabase/pull/46317/commits/4ac6d32b606ae5f5a9a329680c56dec1376681a6) I think this make the code not much difference from the original.

Wdyt?